subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/IncrementalCompilationInitializer.java

```diff
@@ -104,13 +104,17 @@ private void prepareJavaPatterns(Collection<String> staleClasses, PatternSet fil
             String path = staleClass.replaceAll("\\.", "/");
             filesToDelete.include(path.concat(".class"));
             filesToDelete.include(path.concat(".java"));
+            filesToDelete.include(path.concat(".groovy"));
```

Comment:
I think this is much simpler for Groovy: We have the exact mapping which class files have been generated for which source file. And we know the name of the class which has been generated. So what we need to is to delete all the output files which have been created by the source we are going to recompile.

The source files we are going to recompile is directly determined by the stale classes - we have the mapping from class to source file available.

We can at least drop all the inner class patterns with Groovy. I also don't understand how we can have `.java` or `.groovy` classes in the output.

Reply:
See `GroovyRecompilationSpecProvider`.